Initial implementation

mook-as opened the pull request:
This expands upon [f0rmiga/initial-impl](https://github.com/SUSE/eirini-dns-aliases/tree/f0rmiga/initial-impl) to work; this has been tested on a KubeCF deployment.

Notable changes:
- Fixed up the helm chart to actually work. (Unclear why that was a separate chart, but since it was I just kept going.)
- The docker image is now built without the preexisting binary as a dependency (but builds it from source in a separate container instead).  This ensures that we don't end up with things being out of sync (e.g. manually forgetting to rebuild things).
- GitHub Actions workflow for building docker image + helm chart, publishing to GitHub Pages.  Modelled after the pipeline in [eirinix-helm-release](https://github.com/cloudfoundry-incubator/eirinix-helm-release).
- Everything is tagged to include the commit hash.

Note that I don't have write access to the repository, hence not just pushing on top of the branch.  I also can't seem to request a review…

Commit:
feat: initial implementation

mook-as committed:
chore: bump dependencies

In particular, eirinix has moved from github.com/SUSE to
code.cloudfoundry.org.

mook-as committed:
feat: simplify build scripts

mook-as committed:
bump libnetwork

This pulls in newer logrus, which fixes issues with the case of the import
path.  Unfortunately, they hadn't make actual releases in forever, so we
have to pick something off `master`; in this case, the version vendored by
docker 19.03.12.

mook-as committed:
feat: switch to zap for logging in main

Since eirinix.Manager uses zap anyway, use it for main as well to ensure
any logging we do will be in the same format.

mook-as committed:
feat: build binary in docker when building image

This makes us always rebuild the binary when building the image, all in
docker.  This ensures that we have better control over the inputs for that
build.

The base image is still require for running, as (at least for now) the
resulting binary depends on the dynamic linker.

Also includes support to vendor any dependencies if we have any `replace`
lines in `go.mod`; this is useful if we're using a locally edited version
of those dependencies (the most likely of which is eirinix).

mook-as committed:
feat: display binary version on startup

This is useful when developing to ensure the expected binary is in use.

mook-as committed:
feat: add README.

mook-as committed:
feat: scripts to build helm chart

This adds a make target to build the helm chart. Also fixes up the chart to
actually work, including additional role binding (for the webhook
namespace), etc.

mook-as committed:
feat: GitHub Actions support to publish helm chart

This builds the whole thing, publishing docker images and helm charts (the
latter of which goes in GitHub Pages).

mook-as committed:
fix: GitHub Actions: don't run on every push

While running on every push was good for development, we shouldn't do so
for code that we hope will get merged.

viovanov reviewed (changes requested):
Can you please make this a go binary + dockerfile, and move the helm part into kubecf.
We shouldn't change the scope of https://github.com/cloudfoundry-incubator/kubecf/issues/1179

Otherwise 👍

viovanov commented on Dockerfile:
```diff
@@ -0,0 +1,20 @@
+RUN echo "Building eirini-dns-aliases ${version}"
+RUN go build -ldflags="-s -w -X main.appVersion=${version}" -o output/eirini-dns-aliases
+
+FROM "${base}"
```
Have you tried building from scratch?

mook-as replied:
Yes; the binary ended up needing the dynamic linker (`ld-linux.so`) at runtime, and didn't run.

viovanov commented on .github/workflows/publish-helm-chart.yaml:
```diff
@@ -0,0 +1,174 @@
+        echo "GitHub ref found: ${GITHUB_REF}"
+        echo "::set-output name=github-ref::${GITHUB_REF}"
+    - id: git-tag
+      name: Determine Git Tag
```
Please use kubecf-tools versioning script to determine version.

mook-as replied:
I assume that's https://github.com/cloudfoundry-incubator/kubecf-tools/tree/main/versioning ? Is there a readme on what the recommended way to actually use things from that repo is?

viovanov replied:
You can use kubecf as an example [1].
This versioning should only be applied to the docker image now.
We don't have to build another chart (since this is only usable by kubecf), but include the templates directly in kubecf.

https://github.com/cloudfoundry-incubator/kubecf/blob/master/.gitmodules#L4

mook-as reviewed (commented):
I'll redo the helm part in KubeCF — actually I started that way, but saw that the original branch had a helm chart in there.  So it shouldn't be difficult to move back.

mook-as committed:
fix: address review feedback

- Use kubecf-tools to determine the version, instead of `git describe`
  (directly, anyway)
- Drop helm chart handling; assume the user will handle that in some other
  way (i.e. supply their own Kubernetes bits, which is the case for KubeCF)

https://github.com/SUSE/eirini-dns-aliases/pull/1#pullrequestreview-472910075

viovanov commented on .github/workflows/update-helm-index.rb:
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env ruby
```
I think this would be useful in `kubecf-tools`. Please move it there.

mook-as replied:
Removed file completely; found [chart-releaser-action](https://github.com/helm/chart-releaser-action) instead, which looks like it does the same thing (using GitHub Releases for storage), is from upstream helm (instead of a random person), and we don't have to maintain it.  If we _do_ want to keep this, we can just grab it from [eirinix-helm-release](https://github.com/cloudfoundry-incubator/eirinix-helm-release/blob/master/bin/update-helm-index.rb) instead.

I should have removed it already anyway because we moved to only building the image here (and not the chart).

mook-as replied:
Removed in 27470adc887896e5c691becb3f9c5eeedffab05b

viovanov commented on build/manifest.yaml:
```diff
@@ -0,0 +1,12 @@
+# This file controls the arguments shared by the build steps.
```
Please move the binary and image building scripts to kubecf-tools.

mook-as replied:
Will do. Not _quite_ sure yet how they'll work, though, since those are pretty repo-specific.

As a note: the binary (and image building) scripts are unused by CI. But they still make sense to have for local development (mostly), so I'll move them over.

mook-as replied:
Using it from kubecf-tools as of 87814da51cd08e6dfb56171eaf39c34b3ac99489.

mook-as reviewed (commented):
Still need to move build/image scripts to kubecf-tools.

mook-as committed:
fix: CI: remove update-helm-index.rb

This is completely unused now.

mook-as committed:
fix: Build image based on scratch

If we turn off CGO for the binary (which doesn't appear to be needed),
we can get a static binary that then doesn't need the dynamic linker
(and therefore can bse based on an empty image).  Note that we _do_
expect a working /tmp, so we will need to create that in the iamge
first.

mook-as committed:
fix: CI: rename workflow

The workflow only builds the docker image; rename the workflow file to
match that.

mook-as committed:
feat: use kubecf-tools build scripts

f0rmiga commented:
@mook-as I think you may need to adjust this PR to use the new kubecf-apps-dns. It was introduced by https://github.com/cloudfoundry-incubator/kubecf/pull/1313.

mook-as committed:
fix: Use GHCR for image

Use the GitHub container registry instead for image destination.
